### PR TITLE
Test Epoch3.4 and Clarity5

### PR DESCRIPTION
**Pull request by BowTiedRadone**
This PR applies stateless and stateful property-based testing techniques for Epoch 3.4 / Clarity 5 changes.

**Stateless (proptest-rs)**

- `secp256r1-verify`: digest roundtrip, tamper/wrong-key/malformed-key cases, and version-governance checks
- from-consensus-buff?: roundtrip, error/none boundary behavior across epochs, and fuzz stability
- Relay depth filter and epoch gating invariants

**Stateful (madhouse-rs)**

- Relay filter and depth-chain acceptance/rejection at each epoch
- Cross-version Clarity 4/5 call semantics
- SIP-040 Originator/Deny/MaybeSent post-condition sequences

**Infrastructure**

- scenario! macro with deterministic default order + `MADHOUSE=1` randomized mode
- Run: `cargo test -p stackslib --lib chainstate::tests::madhouse`
- Randomized: `MADHOUSE=1 PROPTEST_CASES=50 cargo test -p stackslib --lib chainstate::tests::madhouse`

### CODE REVIEW

File: clarity/src/vm/tests/conversions.rs [hunk 1]
```diff
@@ -770,4 +771,289 @@ proptest! {
             prop_assert_eq!(Value::err_uint(1), evaluation);
         }
     }
+
+    // Empty buffer must return none for every type. Forces the corner case
+    // that random generation would almost never produce.
+    #[test]
+    fn prop_from_consensus_buff_empty_buffer_returns_none(
```
Comment (federico-stacks) [inline clarity/src/vm/tests/conversions.rs:779]: It would be nice to include test tagging for all the `proptest` tests can be easily discovered by CI automations.

```suggestion
    #[tag(t_prop)]
    #[test]
    fn prop_from_consensus_buff_empty_buffer_returns_none(
```

File: stackslib/Cargo.toml [hunk 2]
```diff
@@ -72,6 +72,7 @@ chrono = "0.4.19"
 tempfile = "3.3"
 pinny = { workspace = true }
 proptest = { version = "1.6.0", default-features = false, features = ["std"] }
+madhouse = { git = "https://github.com/stacks-network/madhouse-rs.git", tag = "0.2.1" }
```
Comment (federico-stacks) [inline stackslib/Cargo.toml]: Minor: we are currently referencing two versions of madhouse-rs:
- `0.2.1` in `stackslib`
- `0.2.0` in `stacks-node`

if the two versions are compatibles, It might be worth adding `madhouse-rs` as a workspace dependency and referencing it from the crates that use it, to ensure version consistency.

File: Cargo.lock [hunk 3]
```diff
@@ -2099,6 +2099,14 @@ dependencies = [
  "proptest",
 ]
 
+[[package]]
+name = "madhouse"
+version = "0.2.0"
+source = "git+https://github.com/stacks-network/madhouse-rs.git?tag=0.2.1#5d4d51bedf2d56c8ef0643d891fb085df1615c91"
+dependencies = [
```
Comment (federico-stacks) [inline Cargo.lock:2503]: nit: version reported in [madhouse-rs/Cargo.tml](https://github.com/stacks-network/madhouse-rs/blob/0.2.1/Cargo.toml) is not aligned with tag.

File: stackslib/src/chainstate/tests/madhouse/commands/depth.rs [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+/// Epoch34 max call-stack depth.
+const EPOCH34_DEPTH_LIMIT: usize = 128;
+
+/// Expected return value from a successful ping call: `(ok true)`.
+fn ok_true() -> Value {
+    Value::Response(ResponseData {
+        committed: true,
+        data: Box::new(Value::Bool(true)),
+    })
+}
```
Comment (federico-stacks) [inline stackslib/src/chainstate/tests/madhouse/commands/depth.rs]: nit: this helper is repeated over 3 modules. Maybe it can be moved under `commands/mod.rs`

File: stackslib/src/chainstate/tests/madhouse/commands/relay.rs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+
+        // The AST nesting depth includes wrapper overhead from
+        // `(define-public (deep) (ok ...))`.
+        let ast_depth = (self.depth + WRAPPER_OVERHEAD) as u64;
+        let call_stack_limit = if state.chain_epoch() >= StacksEpochId::Epoch34 {
+            128
+        } else {
+            64
+        };
+        let parser_limit = call_stack_limit + AST_DEPTH_BUFFER;
```
Comment (federico-stacks) [inline stackslib/src/chainstate/tests/madhouse/commands/relay.rs:90]: Could we leverage `StackDepthLimits::for_epoch` here? Then we could drop all the constants: `64`,  `128` and `AST_DEPTH_BUFFER`.

Or is the goal to replicate the stack depth computation to guard against changes in the underlying implementation? (and in this case we could just write some unit tests specifics for `StackDepthLimits`)

File: stackslib/src/chainstate/tests/madhouse/commands/postcond.rs [hunk 6]
```diff
@@ -0,0 +1,518 @@
...
+    }
+
+    fn apply(&self, state: &mut Epoch33ToEpoch34TestState) {
+        // Verify the model agrees with the chain before branching.
+        let chain_epoch = state.chain_epoch();
+        assert_eq!(
+            state.current_epoch, chain_epoch,
+            "CallRestrictWithStxCombinedExceeds: model epoch {:?} disagrees with chain {:?}",
+            state.current_epoch, chain_epoch,
+        );
```
Comment (federico-stacks) [inline stackslib/src/chainstate/tests/madhouse/commands/postcond.rs:246]: sanity-check: I noticed that `CallRestrictWithStxCombinedExceeds` and  `CallAsContractWithStxCombinedExceeds` have epoch check, while `CallRestrictWithStxSafe` and `CallAsContractWithStxSafe` haven't. Should they have it?

File: stackslib/src/chainstate/tests/madhouse/commands/sip040.rs [hunk 7]
```diff
@@ -0,0 +1,845 @@
...
+        // Block 1: Originator mode rejected pre-Epoch34.
+        let tx_originator = ConsensusUtils::new_call_tx_with_postconds(
+            state.next_nonce,
+            "nft",
+            "mint",
+            &[Value::UInt(state.nft_next_id as u128)],
+            TransactionPostConditionMode::Originator,
+            vec![],
+        );
+        state.nft_next_id += 1;
```
Comment (federico-stacks) [inline stackslib/src/chainstate/tests/madhouse/commands/sip040.rs]: nit: this feels a bit inconsistent with other commands. Should the NFT ID be incremented before the transaction is created, as done elsewhere?

File: clarity/src/vm/tests/proptest_utils.rs [hunk 8]
```diff
@@ -824,3 +824,169 @@ pub fn utf8_string_literal(data: &UTF8Data) -> String {
...
+        len_range.clone().prop_map(|n| format!("(buff {n})")),
+        len_range
+            .clone()
+            .prop_map(|n| format!("(string-ascii {n})")),
+        len_range.clone().prop_map(|n| format!("(string-utf8 {n})")),
+        Just("(optional int)".to_string()),
+        len_range.prop_map(|n| format!("(list {n} int)")),
+        Just("(response int int)".to_string()),
+    ]
+    .boxed()
```
Comment (federico-stacks) [inline clarity/src/vm/tests/proptest_utils.rs:992]: sanity-check: should `tuple` also be considered here?

File: clarity/src/vm/tests/epoch_gating.rs [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+use stacks_common::types::StacksEpochId;
+
+use crate::vm::ClarityVersion;
+
+/// `default_for_epoch` must be monotonically non-decreasing: later epochs
+/// never downgrade the default ClarityVersion.
+#[test]
+fn test_default_for_epoch_is_monotonic() {
+    // No Clarity in Epoch10.
+    let clarity_epochs = &StacksEpochId::ALL[1..];
```
Comment (federico-stacks) [inline clarity/src/vm/tests/epoch_gating.rs]: nit: readability could be improved using `StacksEpochId::since()`
```suggestion
    let clarity_epochs = StacksEpochId::since(StacksEpochId::Epoch20);
```

File: stackslib/src/chainstate/tests/madhouse/commands/postcond.rs [hunk 10]
```diff
@@ -0,0 +1,518 @@
...
+use crate::chainstate::tests::consensus::{ConsensusUtils, TestBlock};
+use crate::chainstate::tests::madhouse::context::Epoch33ToEpoch34TestContext;
+use crate::chainstate::tests::madhouse::state::Epoch33ToEpoch34TestState;
+
+fn ok_true() -> Value {
+    Value::Response(ResponseData {
+        committed: true,
+        data: Box::new(Value::Bool(true)),
+    })
+}
```
Comment (brice-stacks) [inline stackslib/src/chainstate/tests/madhouse/commands/postcond.rs]: This function shows up several times. Should it be moved to some shared place instead?

Comment (brice-stacks) [reply stackslib/src/chainstate/tests/madhouse/commands/postcond.rs]: Ah, I missed Federico's comment that already pointed this out. Resolving this one.

File: clarity/src/vm/tests/conversions.rs (hunk 1 above)
Comment (radu-stacks) [reply clarity/src/vm/tests/conversions.rs:779]: 80dd989

File: stackslib/Cargo.toml (hunk 2 above)
Comment (radu-stacks) [reply stackslib/Cargo.toml]: b00f957

File: Cargo.lock (hunk 3 above)
Comment (radu-stacks) [reply Cargo.lock:2503]: Thanks! PR opened upstream: https://github.com/stacks-network/madhouse-rs/pull/33.

File: stackslib/src/chainstate/tests/madhouse/commands/relay.rs (hunk 5 above)
Comment (radu-stacks) [reply stackslib/src/chainstate/tests/madhouse/commands/relay.rs:90]: It's the latter: the goal is to replicate the computation and guard against silent changes in the underlying implementation. If `max_call_stack_depth_for_epoch` or `AST_CALL_STACK_DEPTH_BUFFER` were accidentally modified, tests that delegate to `StackDepthLimits::for_epoch` would be tautological (testing the implementation against itself). The hardcoded values act as independent witnesses of the expected behavior.

Adding dedicated unit tests for `StackDepthLimits` would just duplicate what these commands already cover with the same hardcoded assertions, so I'd rather not add them unless you feel strongly.

File: stackslib/src/chainstate/tests/madhouse/commands/postcond.rs (hunk 6 above)
Comment (radu-stacks) [reply stackslib/src/chainstate/tests/madhouse/commands/postcond.rs:246]: It is intentional. The safe variants expect `(ok true)` regardless of epoch, so there's no epoch-dependent branch that a model/chain mismatch could silently mislead. The exceeds variants branch on `is_epoch34()` (block rejection vs `(err u0)`), so the epoch check is a precondition to ensure the branch is evaluating the right case.

File: stackslib/src/chainstate/tests/madhouse/commands/depth.rs (hunk 4 above)
Comment (radu-stacks) [reply stackslib/src/chainstate/tests/madhouse/commands/depth.rs]: 3e9d08a

File: stackslib/src/chainstate/tests/madhouse/commands/sip040.rs (hunk 7 above)
Comment (radu-stacks) [reply stackslib/src/chainstate/tests/madhouse/commands/sip040.rs]: 114b1d8

File: clarity/src/vm/tests/epoch_gating.rs (hunk 9 above)
Comment (radu-stacks) [reply clarity/src/vm/tests/epoch_gating.rs]: 77dcd07

**Review by brice-stacks — DISMISSED** [review]

File: stackslib/src/chainstate/tests/madhouse/commands/relay.rs (hunk 5 above)
Comment (federico-stacks) [reply stackslib/src/chainstate/tests/madhouse/commands/relay.rs:90]: I see your point about avoiding tautological tests and using hardcoded values as an independent check, that makes sense.

That said, I’m wondering if we might be mixing concerns a bit between unit tests and madhouse tests (just calling them “madhouse” here for distinction, but could be any higher-level test). My thinking is:

We could add focused unit tests for `StackDepthLimits::for_epoch` that explicitly assert the expected values (including the constants). That would lock in the behavior in a single, well-scoped place.
Then, in madhouse tests, we could rely on `StackDepthLimits::for_epoch` directly, since at that point we’re more interested in validating system behavior end-to-end (e.g. `RelayDeepContract`) rather than re-checking the exact computation.

This way:
- The computation is still guarded against accidental changes via dedicated unit tests related to the component that encapsulate the behaviour (which we don’t currently have).
- madhouse tests stay simpler and avoid duplicating logic or constants (and try to cross-reference with the original ones by docstring)
- If the computation ever legitimately changes, we only need to update expectations in one place, avoiding duplication across test layers.
- No duplication between tests because they would be layered (if we dont need additional safeguard).

Do you think this split could work?

File: stackslib/Cargo.toml (hunk 2 above)
Comment (federico-stacks) [reply stackslib/Cargo.toml]: just one more sanity-check: are we fine to referencing dependency by tag instead of commit/revision? The fact that it come from `stacks-network` make the difference?

Comment (federico-stacks) [reply stackslib/Cargo.toml]: I missed this commit: d5a8d74192444867e7268c5f0def98a6c200eb39.

Using rev actually.

File: Cargo.lock (hunk 3 above)
Comment (federico-stacks) [reply Cargo.lock:2503]: Related commit: d5a8d74192444867e7268c5f0def98a6c200eb39

File: stackslib/src/chainstate/tests/madhouse/commands/relay.rs (hunk 5 above)
Comment (radu-stacks) [reply stackslib/src/chainstate/tests/madhouse/commands/relay.rs]: That makes perfect sense! Adding it.

File: clarity/src/vm/tests/proptest_utils.rs (hunk 8 above)
Comment (radu-stacks) [reply clarity/src/vm/tests/proptest_utils.rs]: [b3713e8](https://github.com/stacks-network/stacks-core/pull/6912/commits/b3713e843230bf4f58dee060b3d2badb48c986c7)

File: stackslib/src/chainstate/tests/madhouse/commands/relay.rs (hunk 5 above)
Comment (radu-stacks) [reply stackslib/src/chainstate/tests/madhouse/commands/relay.rs]: [cb15ada](https://github.com/stacks-network/stacks-core/pull/6912/commits/cb15adadd061c9a02fd325eb9ca26f0a46e4cad1)

**Review by brice-stacks — APPROVED** [review]

**Review by federico-stacks — APPROVED** [review]
